### ISSUE TITLE
Callback not being called on File End

### ISSUE THREAD

**Issue by tragicmanner** [MajicDesigns/MD_YX5300#7]
## IMPORTANT
Before submitting this issue
[x] Have you tried using the latest version of the library?
[x] Have you checked this has not already been submitted and/or resolved?
[x] If you are requesting help a better choice may be the [ Arduino forum](http://forum.arduino.cc/)

## Subject of the issue
Callback not being called on File End

## Your Environment
**Library Version:**  1.2.3
**Arduino IDE version:** 1.8.8
**Hardware model/type:** Arduino Nano
**OS and Version:** 

## Steps to Reproduce
Define a callback method that handles STS_FILE_END, set it using setCallback(), play a music file, put in a Serial.print to output "File Ended" when callback is called and the status->code is STS_FILE_END.

## Expected Behaviour
As the file ends, we should see the callback trigger and hit the Serial.print statement

## Actual Behaviour
As the file ends, our Serial Monitor is completely quiet.

## Code Demonstrating the Issue
````
// Callback method for audio library
void cbResponse(const MD_YX5300::cbData *status)
// Callback function used to process device unsolicited messages
// or responses to data requests
{
  Serial.print("callback\n");
  switch (status->code)
  {
    case MD_YX5300::status_t::STS_FILE_END:   // track has ended
      Serial.print("Track Ended\n");
      sfx.playStop();
      playing = false;
      break;
  }
}

...

void setup() {
  Serial.begin(9600);    
  // initialize audio
  sfx.begin();
  sfx.setSynchronous(false);
  sfx.setCallback(cbResponse);
  sfx.volume(30);

...
````

**Comment by tragicmanner** [comment]
So far the library has been great to use, but I am struggling with the asychronous nature of some aspects of the YX5300.

I am wondering if I have initialized things properly and included my code above. If not, please let me know! But so far I have managed to sometimes get the callback to work, but it is very intermittent. I tried a work around where I use getStatus() to determine if the YX5300 is currently playing audio. My program needs to wait for certain audio files to stop playing before moving on to other parts of the program, so it would be preferable to do this by detecting when playing ends and continuing from there.

**Comment by MajicDesigns** [comment]
Have you read the blog posts I put up about the message sequencing, which may help you bridge any gaps you may have:
https://arduinoplusplus.wordpress.com/2018/07/23/yx5300-serial-mp3-player-catalex-module/
https://arduinoplusplus.wordpress.com/2019/04/25/yx5300-serial-mp3-player-message-sequencing/
and this one describing the MP3 player that totally relies on the end of file callback to work properly
https://arduinoplusplus.wordpress.com/2019/05/12/yx5300-mp3-player/

**Comment by tragicmanner** [comment]
Yes, these have been very helpful to get started, but I'm not sure what I am doing wrong. I seem to be doing everything I need to be as far as I can tell? When check() is called, if there are messages to process then callback is triggered from the library, right? Is it possible that other messages have come in and kept the messages I need to process from being processed?

**Comment by MajicDesigns** [comment]
- I would set up for synchronous operations with callback. Synchronous means that you will wait for the ack response on a message pair and eliminate that message from the callback. 
- Make sure that you call check() every time through loop() (can't see that in your code snippet).
- For debugging print out any messages you don't handle in cbResponse (default: switch case) so that you can see what is coming in that you are not explicitely handling.
- Have a look at the MP3 player examples to see if there is any material difference in how they handle this, as they work as far as I can tell.

**Comment by tragicmanner** [comment]
I was trying Synchronous and that really added a lot more processing time than I was anticipating, which really slowed the program down, which is a problem because a lot of the program is involved in the timing of LED animations and things like that.

I do currently have a check() in my loop:
```
void loop() {
  // get current sfx state
  sfx.check();
```
I do not do anything explicitly with the result of the check(), as I'm assuming that registering the callback handles this sort of thing.

The program did have a delay incorporated in the loop, which I removed out of fear of missing a callback.

I also had added a default where I used Serial.print to output anything I was not explicitly handling. Do I need to be doing something to acknowledge the callback to allow for another one to be processed?

**Comment by MajicDesigns** [comment]
> added a lot more processing time

I assume you mean it took longer to return, which would be the case.

> I'm assuming that registering the callback handles this sort of thing

The check() looks right. Yes, you would expect that you get the callbacks when the message arrives with no further action required.

>  acknowledge the callback to allow for another one to be processed

No, it is a 'one way' system. It just tells you they are there.

> output anything I was not explicitly handling

So here you should be seeing all the ACK messages for the non-synchronous commands you are sending. Do you see these as you would expect?

Other things that could be wrong are wiring or hardware related. How good/reliable are the serial connections?

**Comment by tragicmanner** [comment]
I am getting back 0x41 after making calls. Right now I have simplified things for debug purposes so my program simply calls the audio file on start and makes no more calls to the library after that to see if I get back 0x3d, but I am not getting it back. The audio file is only 6 seconds in length, and there are no other callbacks occurring during that time and I do not expect anything other than the ack callback. Simply the 0x41 when starting the device, 0x41 when playing the file. From time to time I will see a 0x03, though I am not sure why.

I have everything hooked up on a breadboard right now, and will swap the jumper cables I am using in case they are a problem. I do have an extra YX5300 I can use to see if the other one is faulty.

**Comment by MajicDesigns** [comment]
0x41 is and ACK_OK. 0x03 is a bad checksum. It may be possible that the message being returned has a bad checksum so you don't get the FILE_END (0x3d) because it is in a 'bad' message. Unlikely that the software ends would suddenly start getting the checksums wrong, but it is possible. Most likely a noisy/loose wire connection. You could try disabling checksums (#define USE_CHECKSUM in the library header) or checking the wiring.

**Comment by tragicmanner** [comment]
Jumper cables did not seem to be the issue, but when I swapped out the YX5300 I did see a slight change in behavior.

With the new YX5300 around the time the audio file ends, I sometimes get back 0x41, I sometimes get back 0x03.

I set USE_CHECKSUM to 0 in the header file and I finally got back 0x3d at the end of the file playing!  I did hit a hiccup, though:

I restarted the device 10 times, and out of those ten times:
2/10 times I got back nothing when the file ended playing
5/10 times I got back 0x3d
3/10 times I got back 0x9e!

I have no idea why I'm getting back 0x9e, or even what that means, but it is very consistent in giving back 0x9e from time to time.

Through all of this I have always gotten back 0x41 after starting the audio file, very consistently, which makes me believe that the connection is reliable enough for that, at least.

**Comment by MajicDesigns** [comment]
0x9e does not seem to be a documented code, so either it is a new code or its is in the bad checksum message (which is why it is 'bad'). This is a strange one, and I don't really have any other really good suggestions at this stage. 
1. If you have a serial line sniffer it could be worth checking the message traffic and verifying that the data you see on the line is the same as what the library receives. This is getting into low level debugging...
2. Maybe leave checksums off and interpret 0x9e (as long as it is consistent) as the end of the file being played as a workaround.
3. If your file is consistently 6 seconds long then implement a timer in the code to 'know' when the file has finished as another workaround.
4. Maybe a combinations of 2 and 3 above so the timer acts as a watchdog for not getting the message back.

**Comment by tragicmanner** [comment]
I have not used a serial line sniffer before, but I'll look into it.

I was thinking of doing that with 0x9e and just assume it's an end of file callback, so I'll probably implement that.

Since all my audio files are static length except for a couple of scenarios where audio files are looped, I'm going to see if I can get things working that way. I actually did two other projects involving a YX5300 where I used my own library to send commands to it, but I did not listen back at all. So that shouldn't be too hard to implement. It's too bad I can't get this to be more reliable, but ultimately you just have to work with what you have :D 

Thanks again for all the help so far. I'll comment if I can figure out anything further as I continue to tinker.
